File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.cpp
#
# @lc app=leetcode id=3598 lang=cpp
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#
# @lc code=start
class Solution {
public:
    int computeLCP(const string& s1, const string& s2) {
        int len = 0;
        int minLen = min(s1.length(), s2.length());
        for (int i = 0; i < minLen; i++) {
            if (s1[i] == s2[i]) {
                len++;
            } else {
                break;
            }
        }
        return len;
    }
    
    vector<int> longestCommonPrefix(vector<string>& words) {
        int n = words.size();
        vector<int> answer(n, 0);
        
        if (n <= 1) {
            return answer;
        }
        
        // Precompute LCP for all adjacent pairs
        vector<int> lcp(n - 1);
        for (int i = 0; i < n - 1; i++) {
            lcp[i] = computeLCP(words[i], words[i + 1]);
        }
        
        // For each removal
        for (int i = 0; i < n; i++) {
            int maxLCP = 0;
            
            // Check all existing pairs that don't involve index i
            for (int j = 0; j < n - 1; j++) {
                if (j != i - 1 && j != i) {
                    maxLCP = max(maxLCP, lcp[j]);
                }
            }
            
            // Check new pair if both neighbors exist
            if (i > 0 && i < n - 1) {
                int newLCP = computeLCP(words[i - 1], words[i + 1]);
                maxLCP = max(maxLCP, newLCP);
            }
            
            answer[i] = maxLCP;
        }
        
        return answer;
    }
};
# @lc code=end

File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.py
#
# @lc app=leetcode id=3598 lang=python3
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#

# @lc code=start
from typing import List

class Solution:
    def longestCommonPrefix(self, words: List[str]) -> List[int]:
        def lcp(a: str, b: str) -> int:
            res = 0
            l = min(len(a), len(b))
            while res < l and a[res] == b[res]:
                res += 1
            return res
        n = len(words)
        answer = [0] * n
        if n <= 1:
            return answer
        prefix_len = [0] * (n - 1)
        for j in range(n - 1):
            prefix_len[j] = lcp(words[j], words[j + 1])
        pre_max = [0] * n
        for j in range(n - 1):
            pre_max[j + 1] = max(pre_max[j], prefix_len[j])
        suf_max = [0] * n
        suf_max[n - 1] = 0
        for j in range(n - 2, -1, -1):
            suf_max[j] = max(suf_max[j + 1], prefix_len[j])
        for i in range(n):
            if i == 0:
                kept_max = 0 if n < 2 else suf_max[1]
            elif i == n - 1:
                kept_max = 0 if n < 2 else pre_max[n - 2]
            else:
                kept_max = max(pre_max[i - 1], suf_max[i + 1])
            new_lcp_val = 0
            if 1 <= i <= n - 2:
                new_lcp_val = lcp(words[i - 1], words[i + 1])
            answer[i] = max(kept_max, new_lcp_val)
        return answer
# @lc code=end